soc/arm/nordic_nrf/soc_nrf_common.h

```diff
@@ -13,6 +13,155 @@
+ *
+ * This macro converts a devicetree GPIO phandle array value
+ * "<&<gpioX> <pin> ...>" to a "PSEL value".
+ *
+ * Note: in Nordic SoC devicetrees, "gpio0" means P0, and "gpio1"
+ * means P1.
+ *
+ * Examples:
+ *
+ *     foo: my-node {
+ *             tx-gpios = <&gpio0 4 ...>;
+ *             rx-gpios = <&gpio1 5 ...>;
+ *     };
+ *
+ *     NRF_DT_GPIOS_TO_PSEL(DT_NODELABEL(foo), tx_gpios) // 0 + 4 = 4
+ *     NRF_DT_GPIOS_TO_PSEL(DT_NODELABEL(foo), rx_gpios) // 32 + 5 = 37
+ */
+#define NRF_DT_GPIOS_TO_PSEL(node_id, prop)				\
+	(DT_GPIO_PIN(node_id, prop) +					\
+	 NRF_PSEL_OFFSET_HELPER(node_id, prop))
+
+/**
+ * Error out the build if the devicetree node with identifier
+ * 'node_id' has both a legacy psel-style property and a gpios
+ * property.
+ *
+ * Otherwise, do nothing.
+ *
+ * @param node_id node identifier
+ * @param psel_prop lowercase-and-underscores PSEL style property
+ * @param psel_prop_name human-readable string name of psel_prop
+ * @param gpios_prop lowercase-and-underscores foo-gpios style property
+ * @param gpio_prop_name human-readable string name of gpios_prop
+ */
+#define NRF_DT_CHECK_PSEL(node_id, psel_prop, psel_prop_name,		\
+			  gpios_prop, gpios_prop_name)			\
+	BUILD_ASSERT(							\
+		(DT_NODE_HAS_PROP(node_id, psel_prop) ^			\
+		 DT_NODE_HAS_PROP(node_id, gpios_prop)),		\
+		"Devicetree node " DT_NODE_PATH(node_id)		\
+		" has neither or both of the " psel_prop_name		\
+		" and " gpios_prop_name					\
+		" properties set; you must set exactly one. "		\
+		"Note: use /delete-property/ to delete properties.")
+
+/**
+ * Error out the build if 'prop' is set on node 'node_id' and
+ * DT_GPIO_CTLR(node_id, prop) is not an SoC GPIO controller.
+ *
+ * Otherwise, do nothing.
+ *
+ * @param node_id node identifier
+ * @param psel_prop lowercase-and-underscores PSEL style property
+ * @param psel_prop_name human-readable string name of psel_prop
+ * @param gpios_prop lowercase-and-underscores foo-gpios style property
+ * @param gpio_prop_name human-readable string name of gpios_prop
+ */
+#define NRF_DT_CHECK_GPIO_CTLR_IS_SOC(node_id, prop, prop_name)		\
+	BUILD_ASSERT(!DT_NODE_HAS_PROP(node_id, prop) ||		\
+		     NRF_DT_GPIO_CTLR_IS_SOC(node_id, prop),		\
+		     "Devicetree node " DT_NODE_PATH(node_id)		\
+		     " property " prop_name				\
+		     " GPIO controller must be &gpio0 "			\
+		     "or &gpio1 (if available on the SoC), but got "	\
+		     DT_NODE_PATH(DT_GPIO_CTLR(node_id, prop))		\
+		     ", which is neither of these")
+
+/*
+ * Returns true if DT_GPIO_CTLR(node_id, prop) is an SoC GPIO
+ * controller (i.e. same node as DT_NODELABEL(gpio0) or
+ * DT_NODELABEL(gpio1)). Otherwise, returns false.
+ */
+#define NRF_DT_GPIO_CTLR_IS_SOC(node_id, prop)				\
+	(NRF_DT_GPIO_CTLR_IS_GPIOX(node_id, prop, 0) ||			\
+	 UTIL_AND(DT_NODE_EXISTS(DT_NODELABEL(gpio1)),			\
+		  NRF_DT_GPIO_CTLR_IS_GPIOX(node_id, prop, 1)))
+
+/* Is DT_GPIO_CTLR(node_id, prop) the same node as DT_NODELABEL(gpio<idx>)? */
```

Comment:
Instead of having a dependency on names, why don't we add a `bank` property to the GPIO node, with 0 for `NRF_P0` and 1 for `NRF_P1` and just multiply that by 32 as an offset?  That's also forward-looking to a potential `NRF_P2`.

Reply:
> Instead of having a dependency on names, why don't we add a `bank` property to the GPIO node, with 0 for `NRF_P0` and 1 for `NRF_P1` and just multiply that by 32 as an offset? That's also forward-looking to a potential `NRF_P2`.

Sure, I could do that if that's the consensus. This was just the easiest thing I could think of that didn't require going into all the .dtsi files.

Another alternative would be to match against `DT_REG_ADDR(node_id)` like `CHECK_DT_REG(gpio0, NRF_P0);` does in validate_base_addresses.c.

I'm not worried about NRF_P2 because that would require a new SoC port, at which point updating this file would be easy enough to do.